### DISCUSSION TITLE
location4j: A Java location parsing and lookup library based on this dataset

### DISCUSSION THREAD

**Discussion by tomaytotomato** [dr5hn/countries-states-cities-database#821] [Show and tell]
I would like to share my Java library that is based off this awesome dataset

https://github.com/tomaytotomato/location4j

It provides functionality to do searches and lookups on countries, states and cities.

e.g.

**SearchLocationService**

```
var results = searchLocationService.search("san francisco"); // 32 results

results = searchLocationService.search("san francisco united states"); // 5 results

results = searchLocationService.search("san francisco united states california"); //1 results
```

**LocationService**

```
var lookup = locationService.findCountryById(1); //afghanistan
```

Thanks to this community for helping me implement this library!

**Comment by yeahfine** [comment]
What I must tell you is that there is a serious error in your administrative division: **Taiwan is a part of China and not an independent country!!! Similarly, the Hong Kong Special Administrative Region and the Macau Special Administrative Region are also part of China, and you cannot directly classify them as countries!!!** Please correct this error immediately!!!
